server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java

```diff
@@ -103,127 +102,75 @@ protected void processTasks(CancelTasksRequest request, Consumer<CancellableTask
         String nodeId = clusterService.localNode().getId();
-        final boolean canceled;
         if (cancellableTask.shouldCancelChildrenOnCancellation()) {
-            DiscoveryNodes childNodes = clusterService.state().nodes();
-            final BanLock banLock = new BanLock(childNodes.getSize(), () -> removeBanOnNodes(cancellableTask, childNodes));
-            canceled = taskManager.cancel(cancellableTask, request.getReason(), banLock::onTaskFinished);
-            if (canceled) {
-                // /In case the task has some child tasks, we need to wait for until ban is set on all nodes
-                logger.trace("cancelling task {} on child nodes", cancellableTask.getId());
-                AtomicInteger responses = new AtomicInteger(childNodes.getSize());
-                List<Exception> failures = new ArrayList<>();
-                setBanOnNodes(request.getReason(), cancellableTask, childNodes, new ActionListener<Void>() {
-                    @Override
-                    public void onResponse(Void aVoid) {
-                        processResponse();
-                    }
-
-                    @Override
-                    public void onFailure(Exception e) {
-                        synchronized (failures) {
-                            failures.add(e);
-                        }
-                        processResponse();
-                    }
-
-                    private void processResponse() {
-                        banLock.onBanSet();
-                        if (responses.decrementAndGet() == 0) {
-                            if (failures.isEmpty() == false) {
-                                IllegalStateException exception = new IllegalStateException("failed to cancel children of the task [" +
-                                    cancellableTask.getId() + "]");
-                                failures.forEach(exception::addSuppressed);
-                                listener.onFailure(exception);
-                            } else {
-                                listener.onResponse(cancellableTask.taskInfo(nodeId, false));
-                            }
-                        }
-                    }
-                });
+            StepListener<Void> completedListener = new StepListener<>();
+            GroupedActionListener<Void> groupedListener = new GroupedActionListener<>(ActionListener.map(completedListener, r -> null), 3);
+            Collection<DiscoveryNode> childrenNodes =
+                taskManager.startBanOnChildrenNodes(cancellableTask.getId(), () -> groupedListener.onResponse(null));
+            boolean cancelled = taskManager.cancel(cancellableTask, request.getReason(), () -> groupedListener.onResponse(null));
+            if (cancelled == false) {
+                throw new IllegalStateException("task with id " + cancellableTask.getId() + " is already cancelled");
```

Comment:
> Another issue I have with this is that this does not play nicely with wait_for_completion. Assume someone else has cancelled the task, and you now want to wait for completion. This gives you a hard exception instead of waiting for all child tasks to complete.

I thought about that scenario but decided not to implement it as I worried about its complexity (we need to keep track of a list of cancellation listeners and childTasksCompleted listeners). Anyway, I will work on it.

Reply:
Let's add a TODO for that part, and consider it in a follow-up

Reply:
The `cancel()` method is a bit trappy. I have removed the return value and made it idempotent in https://github.com/elastic/elasticsearch/pull/54312/commits/9a55fe41951f6bf0704ac5fe3690aa1206f832c5.